### PR TITLE
reworked mod_flag_rollover PR

### PR DESCRIPTION
**Pull request by obra**
This PR is a reworking of #5 

Before merge, I'd squash everything together, but I'd love some san checks.

### CODE REVIEW

**Comment by gedankenexperimenter** [comment]
A couple of the name changes are ones that I think are misleading, dropping information about variables and functions. I'll take a closer look at the structural changes when I get the time (likely tomorrow morning).

**Comment by gedankenexperimenter** [comment]
> `newly_toggled_on_keycode` (to me) reads oddly. Do you have thoughts on other
possibly better names?

Not really. The adverb is terrible, but I couldn't think of a better way to describe what it stores without making the name ridiculously long. I still think it's better than `last_keycode_toggled_on`, which – because significant amounts of the state involved in this part of the code _are_ persistent past the report being cleared – strongly implies to me that it will not be cleared. Nevertheless, that value is deliberately hidden by the anonymous namespace so that it won't be part of the API, so I won't put up a stink about it if you disagree with me about the importance of that bit of information in the name. I've stated the case for my opinion. All I ask is that you consider it before making your final decision.

> Up to now, everything other than "lastKeyReport" was ephemeral.

That's only half true. Plenty of plugins store information that isn't cleared along with the report on every cycle. In fact, I think it's more accurate to say that the ephemeral state of the key report is special, because it takes an extra step to deliberately clear it at the end of the cycle. The whole problem that this PR is trying to fix is caused because the key report is inadequate to store the necessary information, and some of that information _must_ persist in order to produce the output that we want in certain cases.

File: src/kaleidoscope/hid.cpp [hunk 1]
```diff
@@ -51,42 +52,172 @@
 namespace kaleidoscope {
 namespace hid {
 
-void initializeKeyboard() {
-  Keyboard.begin();
-  WITH_BOOTKEYBOARD {
-    BootKeyboard.begin();
-  }
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp:58]: This function wasn't actually deleted. I'm concerned that git records it as being deleted and added later (line 136); this messes with the history a bit. I'm not sure why it ended up like this; it wasn't this way in the original version of this PR.

File: src/kaleidoscope/hid.cpp [hunk 2]
```diff
@@ -51,42 +52,172 @@
...
+
+
+// The functions in this namespace are primarily to solve the problem of rollover from a
+// key with a modifier flag (e.g. `LSHIFT(Key_T)`) to one without (e.g. `Key_H`), which
+// used to result in the mod flag being applied to keys other than the one with the
+// flag. By using `modifier_flag_mask`, we can mask out any modifier flags
+// that aren't attached to modifier keys or keys pressed in the most recent cycle,
+// mitigatingthe rollover problem, and getting the intended `The` instead of `THe`.
+
+// requested_modifiers is bitmask of the modifiers attached to any non-modifier key found to be pressed
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: I don't think the term "mask" is appropriate here. It's a bitfield that will be masked by `modifier_flag_mask`, but is not itself used for filtering.

File: src/kaleidoscope/hid.cpp [hunk 3]
```diff
@@ -51,42 +52,172 @@
...
+// during the most recent cycle. For example, it would include modifiers attached to Key_A, but not 
+// modifiers attached to Key_LeftControl)
+
+static uint8_t requested_modifiers{0};
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: Again, I think the name `requested_modifiers` is not as specific as it ought to be. Yes, it's documented properly, but it would be better if one could correctly infer its meaning from the name in the code where it is used, not solely by reading the comment here.

File: src/kaleidoscope/hid.cpp [hunk 4]
```diff
@@ -51,42 +52,172 @@
...
+
+// last_keycode_toggled_on is the keycode of the key most recently toggled on
+// for this report.  This is set when the key is pressed and cleared after
+// the report is sent.
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: I think more emphasis on the fact that this variable gets cleared (or clobbered, when two are detected in the same cycle) would be helpful.

File: src/kaleidoscope/hid.cpp [hunk 5]
```diff
@@ -51,42 +52,172 @@
...
+  if (flags & RALT_HELD) {
     pressRawKey(Key_RightAlt);
   }
-  if (mappedKey.flags & GUI_HELD) {
+  if (flags & GUI_HELD) {
     pressRawKey(Key_LeftGui);
   }
+}
+
+}
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: I had a comment here indicating that this was the end of the anonymous namespace. I find that very useful to keep from having to search for the matching curly brace.

File: src/kaleidoscope/hid.cpp [hunk 6]
```diff
@@ -51,42 +52,172 @@
...
+
+static uint8_t modifier_flag_mask{0};
+
+
+
+// The functions in this namespace are primarily to solve the problem of rollover from a
+// key with a modifier flag (e.g. `LSHIFT(Key_T)`) to one without (e.g. `Key_H`), which
+// used to result in the mod flag being applied to keys other than the one with the
+// flag. By using `modifier_flag_mask`, we can mask out any modifier flags
+// that aren't attached to modifier keys or keys pressed in the most recent cycle,
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: I would change "pressed" here to "pressed or held", since "pressed" is not well defined (often used to mean "toggled on").

File: src/kaleidoscope/hid.cpp [hunk 7]
```diff
@@ -51,42 +52,172 @@
...
+// mitigatingthe rollover problem, and getting the intended `The` instead of `THe`.
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: s/mitigatingthe/mitigating the/

File: src/kaleidoscope/hid.cpp [hunk 8]
```diff
@@ -51,42 +52,172 @@
...
+  if (isModifierKey(mappedKey)) {
+    pressModifiers(mappedKey.flags);
+  }
+
+  // If, instead, the modifiers are attached to a 'printable' or non-modifier key, we assume that they're
+  // not intended to modify other keys, so we add them to requested_modifiers, and only allow them to
+  // affect the report if the most recent keypress includes those modifiers.
+  else {
+    requestModifiers(mappedKey.flags);
+  }
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: I find the big gap between the `else` and its preceding `if` block unnecessarily difficult to notice, but the code all looks correct to me.

File: src/kaleidoscope/hid.cpp [hunk 9]
```diff
@@ -51,42 +52,172 @@
...
+
+// If the key we're processing is a modifier key, any modifier flags attached to it are added
+// directly to the report along with the modifier from its keycode byte.
+//
+// (A 'modifier key' is one of the eight modifier keys defined by the HID standard:
+// left and right variants of Control, Shift, Alt, and GUI.)
+
+// Eventually it calls pressRawKey.
+
+void pressKey(Key mappedKey, boolean toggledOn) {
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: The style guide specifies underscores for variable names, but since it matches the preexisting `mappedKey` I guess it makes sense to use `toggledOn` instead of `toggled_on`.

File: src/kaleidoscope/hid.cpp [hunk 10]
```diff
@@ -150,15 +282,57 @@ uint8_t getKeyboardLEDs() {
...
   ConsumerControl.sendReport();
 }
 
+
+
+/***************
+ ***************     ConsumerControl
+ ***************/
+
+
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: This block is not related to the changes for this PR.

File: src/kaleidoscope/hid.cpp [hunk 11]
```diff
@@ -171,7 +345,6 @@ void releaseConsumerControl(Key mappedKey) {
   ConsumerControl.release(CONSUMER(mappedKey));
 }
 
-
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: Whitespace unrelated to substantive changes.

File: src/kaleidoscope/hid.cpp [hunk 12]
```diff
@@ -185,7 +358,9 @@ void releaseSystemControl(Key mappedKey) {
 }
 
 
-// Mouse events
+/***************
+ ***************     Mouse
+ ***************/
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/hid.cpp]: This block is not related to the changes for this PR.

**Review by gedankenexperimenter — APPROVED** [review]
All the code changes look correct to me. I have a few nitpicks, but no real objections.

**Comment by gedankenexperimenter** [comment]
> requested_mod_flags -did- stay as requested_modifiers. I don't think that's
wrong though, because at that point how we're storing the maybe-we-will,
maybe-we-won't modifiers is an implementation detail and the fact that they
came in as flags is less relevant than the fact that they've been requested.

What I meant by "information being lost" was the hint about the meaning of the variable in its name, not that there was anything wrong with the code. Sorry for the confusion.

File: src/kaleidoscope/hid.cpp (hunk 11 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: Yes, it'll end up in another commit when I do the squash and final merge.

File: src/kaleidoscope/hid.cpp (hunk 10 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: Yes, it'll end up in another commit when I do the squash and final merge.

File: src/kaleidoscope/hid.cpp (hunk 5 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: Indeed, I remove it because I really, really dislike block closing comments. In my experience, they tend to get out of sync and become lies quite quickly.

That said, I went to the style guide to get myself some backup. At least as of now, I don't have that backup ;) The style guide inherited from Google doesn't agree with me. I'll put it back.

File: src/kaleidoscope/hid.cpp (hunk 1 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp:58]: That's an artifact of how git computes the diff and isn't really related to this actual changes to the file. (And this can change depending on the arguments fed to the diffing algorithm.) I believe this is a non-issue.

File: src/kaleidoscope/hid.cpp (hunk 12 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: Yes, it'll end up in another commit when I do the squash and final merge.

**Comment by obra** [comment]
> > newly_toggled_on_keycode (to me) reads oddly. Do you have thoughts on other
> > possibly better names?
> > 
> 
> Not really. The adverb is terrible, but I couldn't think of a better way to describe what it stores without making the name ridiculously long. I still think it's better than last_keycode_toggled_on, which – because significant amounts of the state involved in this part of the code are persistent past the report being cleared – strongly implies to me that it will not be cleared. Nevertheless, that value is deliberately hidden by the anonymous namespace so that it won't be part of the API, so I won't put up a stink about it if you disagree with me about the importance of that bit of information in the name. I've stated the case for my opinion. All I ask is that you consider it before making your final decision.

Well sure. If I wasn't considering it, I wouldn't have been asking for suggestions of how to make it better. I'd just have merged it and been done with it.

> > Up to now, everything other than "lastKeyReport" was ephemeral.
> 
> 
> That's only half true. Plenty of plugins store information that isn't cleared along with the report on every cycle. In fact, I think it's more accurate to say that the ephemeral state of the key report is special, because it takes an extra step to deliberately clear it at the end of the cycle. 

I'm speaking specifically of Kaleidoscope, not Kaleidoscope plugins. 

> The whole problem that this PR is trying to fix is caused because the key report is inadequate to store the necessary information, and some of that information must persist in order to produce the output that we want in certain cases.

I'm not debating that. What I was trying to say is just that in the kaleidoscope core, the default assumption is that information about the state of the key report is built from scratch on each cycle.
> 

> What I meant by "information being lost" was the hint about the meaning of the variable in its name, not that there was anything wrong with the code. Sorry for the confusion.

Yup. I got that.

File: src/kaleidoscope/hid.cpp (hunk 9 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: Yup. Most important is that they should -match-, but that's not a reason that both of them shouldn't eventually change.

File: src/kaleidoscope/hid.cpp (hunk 8 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: It's not unreasonable to move the comment inside the block. Thanks.

File: src/kaleidoscope/hid.cpp (hunk 5 above)
Comment (gedankenexperimenter) [reply src/kaleidoscope/hid.cpp]: I've started to switch from `astyle` to `clang-format`, and one of the things it does (at least with the Google style option) is add (and rewrite) comments that close namespaces, so if it gets run before commits, it would prevent those comments from getting stale.

**Comment by gedankenexperimenter** [comment]
> What I was trying to say is just that in the kaleidoscope core, the default assumption is that information about the state of the key report is built from scratch on each cycle.

I can certainly see why my solution was counter to your intuition. The persistence of those bits of state seemed so obvious to me that I neglected to call it out explicitly in the docs. Sorry about that.

**Comment by obra** [comment]
that’s pretty cool. If somebody, and I’m not deputizing you, soffit to put together a clang format invocation to match our style I’d be thrilled to take it for a test run

File: src/kaleidoscope/hid.cpp (hunk 1 above)
Comment (gedankenexperimenter) [reply src/kaleidoscope/hid.cpp:58]: That's what I expected, but it's still quite surprising. I'm really looking forward to pijul replacing git…

**Comment by obra** [comment]
Since I was already cleaning up style, I took the opportunity to rename function parameters to match the style guide.

At this point, I think this is basically ready to go. I've spent time trying to come up with better names for the variables that @gedankenexperimenter and I disagree about, but haven't found anything I prefer to what I'm using now. Since these are all pretty tightly scoped to this one file, it'll be relatively straightforward to change them if we do figure something better out.

@algernon I'd love it if you could review and put this through its paces.

File: src/kaleidoscope/hid.cpp [hunk 13]
```diff
@@ -51,93 +51,225 @@
...
+
+void initializeKeyboard() {
+  Keyboard.begin();
+  WITH_BOOTKEYBOARD {
+    BootKeyboard.begin();
+  }
+}
+
+// pressKey takes a Key, as well as optional boolean 'toggledOn' which defaults
+// to 'true'
```
Comment (domenzain) [inline src/kaleidoscope/hid.cpp:150]: Does this match the implementation?

File: src/kaleidoscope/hid.cpp [hunk 14]
```diff
@@ -51,93 +51,225 @@
...
+// If the key we're processing is a modifier key, any modifier flags attached to
+// it are added directly to the report along with the modifier from its keycode
+// byte.
+//
+// (A 'modifier key' is one of the eight modifier keys defined by the HID
+// standard: left and right variants of Control, Shift, Alt, and GUI.)
+
+// Eventually it calls pressRawKey.
+
+void pressKey(Key pressed_key, boolean toggled_on) {
```
Comment (domenzain) [inline src/kaleidoscope/hid.cpp:167]:
```cpp
void pressKey(Key pressed_key, boolean toggled_on=true) {
```

File: src/kaleidoscope/hid.cpp (hunk 13 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp:150]: The definition of the API is over here: https://github.com/keyboardio/Kaleidoscope/pull/347/files#diff-dfa257efaf4a9a7ccbb4b63d30e4ba54R12 

I -think- it makes more sense to set the default there, but maybe a more experienced C++ dev can set me straight if I'm wrong.

File: src/kaleidoscope/hid.cpp (hunk 14 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp:167]: (See my comment on the other related change. We set this in the API definition)

**Comment by domenzain** [comment]
My bad, I had missed the matching PR for the API.

Is there some documentation for how the different repos depend on each
other?

File: src/kaleidoscope/hid.cpp (hunk 7 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: thanks. got it.

File: src/kaleidoscope/hid.cpp (hunk 6 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: *nod* got it.

File: src/kaleidoscope/hid.cpp (hunk 3 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: While I don't agree, I also don't feel strongly and have changed it to requested_modiifer_flags

File: src/kaleidoscope/hid.cpp (hunk 2 above)
Comment (obra) [reply src/kaleidoscope/hid.cpp]: Changed to 'bitmap'
 thanks.

**Comment by obra** [comment]
Ok. Having looked at this with fresh eyes after -my- offline time this summer, I changed a name and a bit of doc a little bit closer to @gedankenexperimenter's preferences.

While none of this is really the last word, I'm going to merge now, since this is meaningfully much better than the status quo.

@gedankenexperimenter thanks so much for putting up with my pedantry and tardiness on getting this merged (to say nothing of the work you did to actually, you know, fix the problem.)

**Comment by hoschi** [comment]
With this patch applied I have now a bug when using it with OneShot. [This key](https://github.com/hoschi/Model01-Firmware/blob/my-changes/Model01-Firmware.ino#L136) toggles between held and released state while held, e.g. the underlined menus in an application toggle all the time. [This key](https://github.com/hoschi/Model01-Firmware/blob/my-changes/Model01-Firmware.ino#L129) still works as normal.

**Comment by domenzain** [comment]
I see all OneShot modifiers alternating between pressed and released when held. No normal behaviour for any particular OneShot modifier key.

**Comment by jamadagni** [comment]
OK please advise which library to file the OneShot related issue under, here or Kaleidoscope or OneShot itself…

**Comment by gedankenexperimenter** [comment]
> OK please advise which library to file the OneShot related issue under, here or Kaleidoscope or OneShot itself…

I would submit it as a OneShot issue. Even if the fix belongs in some other module, that's the best place for the issue, because it's related to that plugin.

**Comment by obra** [comment]
Except the same problem is happening to Macros and topsy turvy.  I'd submit
it against the hidadaptor for now.
ᐧ

**Comment by obra** [comment]
@gedankenexperimenter #12 appears to correct the behavior of the new codepath in many (but not all) situations. 

The change in  question is to only set last_keycode_toggled_on if that keycode is -not- a modifier. Based on my understanding of your design, that ought to be correct. Is it desperately wrong in some way I'm not considering?

(The bad behavior was a lot easier to see once I cleaned up some other bits of KeyboardioHID.)

**Comment by gedankenexperimenter** [comment]
That sounds correct to me.

Having taken a look at the OneShot code earlier today, though, I still think it's doing something wrong. It appears to be sending toggled-on (`IS_PRESSED`) events every cycle until cancelled, instead of held (`IS_PRESSED | WAS_PRESSED`). I haven't had enough time to examine everything thoroughly yet, but the OneShot problem didn't seem related to the Macros problem. Can you explain what was going on?

**Comment by obra** [comment]
Without the new change, I was definitely seeing oneshot causing a toggle on
/ toggle off every cycle with the new hidadaptor code. Once I disabled the
code that explicitly turns off the toggled on key, sends a report and then
turns it back on, it got better.

(Due to unrelated inefficiencies in KeyboardioHID, all of these got sent
three times until today's improvements. That made debugging harder but
didn't actually change behavior.)

-j
ᐧ

**Comment by algernon** [comment]
> instead of held `(IS_PRESSED | WAS_PRESSED)`

Nice catch, thank you! I'll keep this in mind next time I get around to hack on OneShot.
